### DISCUSSION TITLE
Ownership and lifetime problem in clap_input_events

### DISCUSSION THREAD

**Discussion by Arakula** [free-audio/clap#246] [General]
As I'm currently implementing CLAP support in my hosts, I came across what I'd consider to be a conceptual problem in clap_input_events:
```
   // Don't free the returned event, it belongs to the list
   const clap_event_header_t *(CLAP_ABI *get)(const struct clap_input_events *list, uint32_t index);
```
The PlugIn expects to receive a pointer to one of the events from the host.
Since the PlugIn obviously can't claim ownership over the returned pointer, which is also more or less documented in the comment above the declaration of `get`, the host is responsible for keeping that pointer alive ... but for how long?

Because, you see ... not all hosts are based on CLAP. Mine aren't; they have an internal event list - and associated internal structures - and convert them to CLAP format on the fly when the PlugIn calls `get`. Currently, I maintain a ring buffer for roughly 10 CLAP-converted events, so it's highly unlikely that an event gets overwritten while the PlugIn might still be interested in its contents, but:

It is unspecified how long that event pointer and the associated content should be kept alive. A wonderful base for misunderstandings and spurious, not easily reproducible "sudden death" scenarios. This should at least be documented.

**Comment by abique** [comment]
The events coming from the host are valid pointers for the whole call to `process()`, after that they're not.
If you need to keep them for a longer period of time, then you have to make a copy.

**Reply by baconpaul** [reply]
we should add "but is only valid for the life of the process call" to the header comment. (I agree that is the right answer).

**Reply by Arakula** [reply]
"must stay valid during the call to process" would be better IMO.

My first implementation, as roughly outlined above, didn't necessarily keep them alive that long. My current one does now (prepares all CLAP events before calling process() - reserves more memory that way, but so be it ...).

**Comment by baconpaul** [comment]
@abique can you moderate / block that user? I don't have perms to do so in this repo.
